### PR TITLE
Upgrade TF azurerm v2 and NatGateway pubIP migration

### PR DESCRIPTION
**Pull request by dkistner**
**How to categorize this PR?**

/area open-source
/kind technical-debt
/priority normal
/platform azure

**What this PR does / why we need it**:
This PR contains migration logic to switch the `azurerm` Terraform provider version from `v1.x.x` to version `v2.x.x` (major version change).

In detail the following changes in the Terraform are required:
- Add `features` field to the Terraform manifest
- Remove the route table and network security group references from the `azurerm_subnet` resource. There are meanwhile associations resources in place, but we need to keep the resource references before we can migrate. Before this migration can be applied all Shoots `Infrastructure` need to be reconciled by a Gardener Azure extension which contains this PR: https://github.com/gardener/gardener-extension-provider-azure/pull/161 (shipped with `v1.14.0`)
- The `azurerm_subnet` resource field `address_prefix` has been moved to `address_prefixes`
- Added the migration steps for the Gardener managed NAT IP [proposed](https://github.com/gardener/gardener-extension-provider-azure/pull/54#issuecomment-664593585) by @ialidzhikov. The `TODO(natipmigration)` changes can be removed in future release when all Shoot `Infrastructure` are reconciled with this change. These changes are required to unblock also https://github.com/gardener/gardener-extension-provider-azure/pull/54

This PR can only be merged when a new Terraformer with this PR https://github.com/gardener/terraformer/pull/54 is in.

**Special notes for your reviewer**:
Create an `Infrastructure` with NatGateway (with an Azure extension not including this PR and an `azurerm` tf provider of version `v1.x.x`). Then reconcile the existing `Infrastructure` an Azure extension which contains these changes.

This Terraformer image can be used for testing: `dominickistner/terraformer:azurerm-2360`

**Release note**:

```improvement operator
The Gardener Azure provider extension is now using the `azurerm` Terraform provider in version `v2.x.x`
```

```action operator
Operators need to ensure that all Azure Shoot `Infrastructure` resource are reconciled with the Gardener Azure provider extension `v1.14.0` (contains a PR which is a prerequisite for this Terraform migration https://github.com/gardener/gardener-extension-provider-azure/pull/161) before applying this version. Please have a special look on hibernated clusters as their `Infrastructure` is might not reconciled for a while.
```

``` improvement operator github.com/gardener/terraformer #54 @dkistner
Terraformer uses now the azurerm provider in version v2.36.0
```

/invite @ialidzhikov 
/invite @kon-angelo

### CODE REVIEW

**Comment by dkistner** [comment]
@ialidzhikov Could you have a look? I will remove the draft of this pr https://github.com/gardener/terraformer/pull/54 when we have reviewed this one here.

**Comment by dkistner** [comment]
> toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit

Concourse seems to pull to much from dockerhub

**Review by ialidzhikov — COMMENTED** [review]
I played with it locally and the migration worked well. Looks good generally, thank you! Minor nits inline.

> Before this migration can be applied all Shoots Infrastructure need to be reconciled by a Gardener Azure extension which contains this PR: #161 (shipped with v1.14.0)

There is one corner case that Infrastructures are no longer reconciled for hibernated Shoot - see https://github.com/gardener/gardener/pull/2258. So technically there might be an Infrastructure of hibernated Shoot that is not reconciled by provider-azure@v1.14.0.

**Comment by ialidzhikov** [comment]
We can also run the infrastructure integration test when there is a release of terraformer. Or alternatively you can also update `chart/images.yaml` with terraformer `v1.5.0-dev-b85b19c9d8ce07882aa1b382e26e4560e04b67cc` (I am not sure whether the component descriptor will like this)?

File: pkg/internal/infrastructure/terraform.go [hunk 1]
```diff
@@ -418,3 +438,25 @@ func findDomainCounts(cluster *controller.Cluster, infra *extensionsv1alpha1.Inf
 		updateDomains: *updateDomainCount,
 	}, nil
 }
+
+// requireNatGatewayIpMigration checks if the Gardener managed NatGateway public ip needs to be migrated.
+// TODO(natipmigration) This can be removed in future versions when the ip migration has been completed.
+func requireNatGatewayIpMigration(infra *extensionsv1alpha1.Infrastructure, config *api.InfrastructureConfig) (bool, error) {
```
[the remark this answers is not in the document]
Comment (dkistner) [reply pkg/internal/infrastructure/terraform.go]: done.

File: pkg/internal/infrastructure/terraform.go [hunk 2]
```diff
@@ -227,10 +235,13 @@ type TerraformState struct {
...
 	// IdentityClientID is the client id of the identity.
 	IdentityClientID string
+	// NatGatewayIPMigrated is the indicator if the nat gateway ip is migrated.
+	// TODO(natipmigration) This can be removed in future versions when the ip migration has been completed.
+	NatGatewayIPMigrated string
 }
 
 // ExtractTerraformState extracts the TerraformState from the given Terraformer.
-func ExtractTerraformState(tf terraformer.Terraformer, config *api.InfrastructureConfig) (*TerraformState, error) {
+func ExtractTerraformState(tf terraformer.Terraformer, infra *extensionsv1alpha1.Infrastructure, config *api.InfrastructureConfig) (*TerraformState, error) {
```
[the remark this answers is not in the document]
Comment (dkistner) [reply pkg/internal/infrastructure/terraform.go]: removed it. Sorry this was a left over.

File: pkg/internal/infrastructure/terraform.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dkistner) [reply pkg/internal/infrastructure/terraform.go]: yep, added some tests for this.

**Comment by dkistner** [comment]
I run the infra integration tests locally with my custom terraformer `dominickistner/terraformer:azurerm-2360` before I opened the pr. Those succeed.
Btw: Addressed also the requested changes. PTAL.

**Comment by ialidzhikov** [comment]
/lgtm


> > Before this migration can be applied all Shoots Infrastructure need to be reconciled by a Gardener Azure extension which contains this PR: #161 (shipped with v1.14.0)
> 
> There is one corner case that Infrastructures are no longer reconciled for hibernated Shoot - see https://github.com/gardener/gardener/pull/2258. So technically there might be an Infrastructure of hibernated Shoot that is not reconciled by provider-azure@v1.14.0.

@dkistner , what is our last agreement about this? Manual detection of such clusters and manual trigger for reconcile? If that is the case, will you take care about this?

**Comment by ialidzhikov** [comment]
You can also switch the base image to eu.gcr.io/gardener-project/3rd/golang:1.15.3 to workaround the dockerhub rate limit issues

**Comment by ialidzhikov** [comment]
> You can also switch the base image to eu.gcr.io/gardener-project/3rd/golang:1.15.3 to workaround the dockerhub rate limit issues

It looks like it breaks the release script - ref https://github.com/gardener/gardener-extension-provider-azure/blob/v1.14.0/vendor/github.com/gardener/gardener/hack/.ci/prepare_release#L38

**Comment by kon-angelo** [comment]
> You can also switch the base image to eu.gcr.io/gardener-project/3rd/golang:1.15.3 to workaround the dockerhub rate limit issues

Better not now since there is the issue with the release script, but we will change the image version along with the revendoring.

File: pkg/internal/infrastructure/terraform_test.go [hunk 3]
```diff
@@ -330,6 +331,50 @@ var _ = Describe("Terraform", func() {
...
+					values, err := ComputeTerraformerChartValues(infra, clientAuth, config, cluster)
+					Expect(err).To(Not(HaveOccurred()))
+					Expect(values).To(BeEquivalentTo(expectedValues))
+				})
+
+				It("should not migrate the NatGateway IP as it is already migrated", func() {
+					infrastructureStatus := api.InfrastructureStatus{
+						NatGatewayPublicIPMigrated: true,
+					}
+					infrastructureStatusMarsheled, err := json.Marshal(infrastructureStatus)
```
Comment (kon-angelo) [inline pkg/internal/infrastructure/terraform_test.go:365]: optional typo fix

**Review by kon-angelo — APPROVED** [review]
/lgtm

**Review by kon-angelo — DISMISSED** [review]

**Comment by kon-angelo** [comment]
docker's rate limit strikes again

**Comment by ialidzhikov** [comment]
> docker's rate limit strikes again

Let's switch the base image to `eu.gcr.io/gardener-project/3rd/golang:1.15.3`. With the g/g@v1.13 vendoring, we will have https://github.com/gardener/gardener/pull/3189 in place.

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Review by kon-angelo — APPROVED** [review]
/lgtm

**Pull request merged**